cpu/lm4f120/periph/timer.c

```diff
@@ -27,132 +29,384 @@
-        WTIMER0_ICR_R  = 0x00000001;                        /* clear timer0A timeout flag */
-        WTIMER0_IMR_R |= 0x00000001;                        /* arm timeout interrupt */
-        ROM_IntPrioritySet(INT_WTIMER0A, 32);
-        timer_irq_enable(dev);
-        timer_start(dev);
-        DEBUG("startTimeout Value=0x%lx\n", ROM_TimerValueGet(WTIMER0_BASE, TIMER_A));
-        return 1;
-    }
-    return -1;
+    if (dev >= TIMER_NUMOF){
+        return -1;
+    }
+
+    config[dev].cb = callback;                          /* User Function */
+    config[dev].diviser = us_per_tick * ROM_SysCtlClockGet()/1000000;
+
+    unsigned int sysctl_timer;
+    unsigned int timer_base;
+    unsigned int timer_side = TIMER_A;
+    unsigned int timer_cfg = TIMER_CFG_SPLIT_PAIR | TIMER_CFG_A_PERIODIC_UP;
+    unsigned int timer_max_val;
+    unsigned int timer_intbit = TIMER_TIMA_TIMEOUT | TIMER_TIMA_MATCH;
+
+    switch(dev){
+#if TIMER_0_EN
+    case TIMER_0:
+        sysctl_timer = SYSCTL_PERIPH_WTIMER0;
+        timer_base = WTIMER0_BASE;
+        timer_max_val = TIMER_0_MAX_VALUE;
+        break;
+#endif
+#if TIMER_1_EN
+    case TIMER_1:
+        sysctl_timer = SYSCTL_PERIPH_WTIMER1;
+        timer_base = WTIMER1_BASE;
+        timer_max_val = TIMER_1_MAX_VALUE;
+        break;
+#endif
+    default:
+        return -1; /* unreachable */
+    }
+
+    if (timer_side == TIMER_A) {
+        timer_cfg |= TIMER_TAMR_TAMIE;
+    }
+    else if (timer_side == TIMER_B) {
+        timer_cfg |= TIMER_TBMR_TBMIE;
+    }
+
+    ROM_SysCtlPeripheralEnable(sysctl_timer);
+
+    ROM_TimerDisable(timer_base, timer_side);
+    ROM_TimerConfigure(timer_base, timer_cfg);
+
+    unsigned long long lltimer_val_max = _scaled_to_ll_value(timer_max_val, config[dev].diviser);
+
+    ROM_TimerPrescaleSet(timer_base, timer_side, lltimer_val_max >> 32);
+    ROM_TimerLoadSet(timer_base, timer_side, lltimer_val_max & 0xFFFFFFFF);
+    ROM_TimerIntClear(timer_base, timer_intbit);
+
+    ROM_TimerIntEnable(timer_base, timer_intbit);
+
+    timer_irq_enable(dev);
+    timer_start(dev);
+
+    return 1;
 }
 
 int timer_set(tim_t dev, int channel, unsigned int timeout)
 {
-    if (dev == TIMER_0) {
-        unsigned int now = timer_read(dev);
-        DEBUG("timer_set now=0x%x\n",now);
-        DEBUG("timer_set timeout=0x%x\n", timeout);
-        return timer_set_absolute(dev, channel, now+timeout);
+    unsigned int correct_now;
+    unsigned long long uncorrected_now;
+    int r;
+
+    if (dev >= TIMER_NUMOF){
+        return -1;
     }
-    return -1;
+
+    switch(dev){
+#if TIMER_0_EN
+    case TIMER_0:
+        break;
+#endif
+#if TIMER_1_EN
+    case TIMER_1:
+        break;
+#endif
+    default:
+        return -1; /* unreachable */
+    }
+
+    correct_now = timer_read(dev);
+    uncorrected_now = _scaled_to_ll_value(correct_now, config[dev].diviser);
+    r = timer_set_absolute(dev, channel, uncorrected_now+timeout);
```

Comment:
`uncorrected_now+timeout` will be multiplied once more in `_scaled_to_ll_value` in `timer_set_absolute`. Does this still provide correct values?